LFRic Apps apply upgrade macros script (Part 2)

james-bruten-mo opened the pull request:
From Issue #37 and building on #38.

Building on the initial script added in #38, this PR adds the code to run the rose commands to apply the new macro and validate the metadata. It also adds unit tests for the entire script (these should be set up for CI but I think we should leave that to it's own ticket and have a coherent setup for the repo).

See #38 for additional links and instructions on testing.

james-bruten-mo commented:
This is blocked by the first part, but may as well go into review in parallel

cameronbateman-mo reviewed (approved):
All looks good, here is my seal of approval

![image](https://github.com/user-attachments/assets/ac59896a-36e0-47be-90dc-00d8eae801fe)

james-bruten-mo commented:
@ericaneininger This ticket is now ublocked - I've fixed merge conflicts from the changes in the part1 branch and updated unit testing where required. Have also run the script on my example working copies - this works as expected.

james-bruten-mo commented:
And I've actually merged in main now instead of manually fixing the conflicts, so the diff appears nicely - getting used to this github lark

ericaneininger reviewed (commented):
These new unit tests do not cover any of the new code being added - they only appear to cover existing code, and I'm not sure how much of that.  Is there an intention to cover all of the code in future?

james-bruten-mo commented on lfric_macros/apply_macros.py:
```diff
@@ -785,6 +785,129 @@ def preprocess_macros(self):
+                continue
+            if app in apps_lstdir:
+                app_path = os.path.join(app_dir_apps, app)
+            else:
```
Have modified this section - I need both the app name and the app path, so have modified your suggestion, but gets rid of the if-else you mention.

ericaneininger replied:
Thanks.  I dd spot you needed both, which is why I suggested adding `apath, aname = os.path.split(app_path)`.

james-bruten-mo commented:
> These new unit tests do not cover any of the new code being added - they only appear to cover existing code, and I'm not sure how much of that. Is there an intention to cover all of the code in future?

Thanks Erica. I don't know whether any of the new code added here is particularly suitable to unit tests as it's mostly applying rose commands -  if you've got a suggestion for how to unit test that then I'm all ears.

ericaneininger reviewed (approved):
Thanks for changes as requested.  Happy to approve.

cameronbateman-mo reviewed (commented):
All looks good apart from expanding out the list comprehension to multiple lines, almost defeats the purpose of it. I think I missed this on the first look.

james-bruten-mo commented on lfric_macros/apply_macros.py:
```diff
@@ -853,13 +852,16 @@ def apps_to_upgrade(self):
         app_dir_core = os.path.join(self.core_source, "rose-stem", "app")
-        apps_list = [os.path.join(app_dir_apps, f) for f in os.listdir(app_dir_apps)]
-        apps_list += [os.path.join(app_dir_core, f) for f in os.listdir(app_dir_core)]
+        apps_list = [
+            os.path.join(app_dir_apps, f) for f in os.listdir(app_dir_apps)
+        ]
+        apps_list += [
+            os.path.join(app_dir_core, f) for f in os.listdir(app_dir_core)
+        ]
```
This is the style used by black when the line length is too long. I actually quite like it, although you're right that in this case it doesn't make too much difference.

Thanks for having a look, but you've also missed the boat on this PR as it's already merged!
